src/Microsoft.AspNet.SignalR.Client/Connection.cs

```diff
@@ -469,39 +473,80 @@ protected virtual string OnSending()
-                                _totalTransportConnectTimeout = TransportConnectTimeout + TimeSpan.FromSeconds(negotiationResponse.TransportConnectTimeout);
+                // Default the beat interval to be 5 seconds in case keep alive is disabled.
+                var beatInterval = TimeSpan.FromSeconds(5);
 
-                                // Default the beat interval to be 5 seconds in case keep alive is disabled.
-                                var beatInterval = TimeSpan.FromSeconds(5);
+                // If we have a keep alive
+                if (negotiationResponse.KeepAliveTimeout != null)
+                {
+                    _keepAliveData = new KeepAliveData(TimeSpan.FromSeconds(negotiationResponse.KeepAliveTimeout.Value));
+                    _reconnectWindow = _disconnectTimeout + _keepAliveData.Timeout;
 
-                                // If we have a keep alive
-                                if (negotiationResponse.KeepAliveTimeout != null)
-                                {
-                                    _keepAliveData = new KeepAliveData(TimeSpan.FromSeconds(negotiationResponse.KeepAliveTimeout.Value));
-                                    _reconnectWindow = _disconnectTimeout + _keepAliveData.Timeout;
+                    beatInterval = _keepAliveData.CheckInterval;
+                }
+                else
+                {
+                    _reconnectWindow = _disconnectTimeout;
+                }
+
+                Monitor = new HeartbeatMonitor(this, _stateLock, beatInterval);
+
+                return StartTransport();
+            }
 
-                                    beatInterval = _keepAliveData.CheckInterval;
-                                }
-                                else
+            Task StartNegotiation()
+            {
+                return transport.Negotiate(this, _connectionData)
+                                .Then(negotiationResponse =>
                                 {
-                                    _reconnectWindow = _disconnectTimeout;
-                                }
+                                    VerifyProtocolVersion(negotiationResponse.ProtocolVersion);
+
+                                    if (negotiationResponse.ProtocolVersion.Equals("2.0") && !string.IsNullOrEmpty(negotiationResponse.RedirectUrl))
+                                    {
+                                        // Ensure the URL ends with a trailing slash.
+                                        if (!negotiationResponse.RedirectUrl.EndsWith("/"))
+                                        {
+                                            negotiationResponse.RedirectUrl += "/";
+                                        }
+
+                                        // Update the URL based on the redirect response and restart the negotiation
+                                        Url = negotiationResponse.RedirectUrl.ToString();
```

Comment:
~~`ToString`~~?

Reply:
Whoops! Good catch. I had this factored to use `System.Uri` to "normalize" the URL but decided to just do the `.EndWith` thing above instead and forgot to remove this.